build_tools/cmake/iree_cross_compile.cmake

```diff
@@ -0,0 +1,219 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+include(iree_macros)
+
+# iree_create_configuration
+#
+# Creates custom commands and targets for an IREE configuration. An IREE
+# configuration means a new IREE CMake invocation with its own set of
+# parameters.
+#
+# This function defines two custom target, `iree_prepare_${config_name}_dir`
+# and `iree_configure_${config_name}`, to drive the creation of a directory
+# for hosting the new IREE configuration and its corresponding `CMakeCache.txt`
+# file, respectively. Callers can then depend on either the file or the
+# target to make sure the configuration is invoked as a dependency.
+#
+# This function is typically useful when cross-compiling towards another
+# architecture. For example, when cross-compiling towards Android, we need
+# to have certain tools first compiled on the host so that we can use them
+# to programmatically generate some source code to be compiled together
+# with other checked-in source code. Those host tools will be generated
+# by another CMake invocation configured by this function.
+#
+# Supported CMake options:
+# - IREE_<config_name>_BINARY_ROOT: the root directory for containing IREE build
+#   artifacts for the given `config_name`. If not specified in caller, this is
+#   set to a directory named as `config_name` under the current CMake binary
+#   directory.
+# - IREE_<config_name>_C_COMPILER: C compiler for the given `config_name`.
+#   This must be defined by the caller.
+# - IREE_<config_name>_CXX_COMPILER: C++ compiler for the given `config_name`.
+#   This must be defined by the caller.
+# - IREE_<config_name>_<option>: switch for the given `option` specifically for
+#   `config_name`. If missing, default to OFF for bool options; default to
+#   IREE_<option> for non-bool variables.
```

Comment:
Why not always default to `IREE_<option>`? That seems more consistent and simpler.

Reply:
Lots of the options aren't applicable to different configurarations when cross-compiling. So I think it'd be easier to explicitly require users to turn things on instead of explicitly disabling things. (And so we don't see ripple effect when changing the default for some optional features.)